### DISCUSSION TITLE
.NET Core 3.1 GC problem (.NET Core not performing GC when low on memory)

### DISCUSSION THREAD

**Discussion by JMan7777** [dotnet/runtime#47659] [Q&A] [answered]
Hello,

I have a bit of a tricky problem.

I have a Linux web app running on Azure (B4 instance) under .NET Core 3.1.11 which does also perform big batch processing jobs.
So I have a lot of large objects and basically iterating through the objects during the processing.

I can see that the memory usage keeps going up during that processing (and down when an automatic GC kicks in) 
but finally comes to a point where the web app restarts as it was exceeding its memory limits.

Now the funny thing is, that when I place a manual GC in the processing loop (let's say perform a GC every 10th large object processing), the memory usage tops out stable around 80% and not even comes close to 100%.

I do the manual GC as follows:

```
GCSettings.LargeObjectHeapCompactionMode = GCLargeObjectHeapCompactionMode.CompactOnce;
GC.Collect(2, GCCollectionMode.Forced);
```

and my GC settings are:

```
    <TieredCompilation>true</TieredCompilation>
    <TieredCompilationQuickJit>false</TieredCompilationQuickJit>
    <TieredCompilationQuickJitForLoops>false</TieredCompilationQuickJitForLoops>
    <RetainVMGarbageCollection>true</RetainVMGarbageCollection>
    <ServerGarbageCollection>true</ServerGarbageCollection>
    <ConcurrentGarbageCollection>true</ConcurrentGarbageCollection>
```

Any idea how I can fine-tune the GC settings that I can get rid of the manual GC's.

Thx a lot.

**Comment by huoyaoyuan** [comment]
It sounds like caused by LOH fragmentation. I'm not sure how the runtime is configured for that, but manually require compaction of LOH should be the right way.

**Reply by huoyaoyuan** [reply]
Continuously setting `LargeObjectHeapCompactionMode` without force GC should be OK. By documentation, it means that LOH will be required to be compacted during next GC.

**Reply by JMan7777** [reply]
> 
> 
> Continuously setting `LargeObjectHeapCompactionMode` without force GC should be OK. By documentation, it means that LOH will be required to be compacted during next GC.

Thx a lot.

I gave it a try but it didn't work so well.

### **Test settings:**

### Test A:

1) One time setting at startup:~~
```
GCSettings.LatencyMode = GCLatencyMode.SustainedLowLatency;
```
This is to reduce full blocking Gen2 GC's and only run them if really needed.
(Enables garbage collection that tries to minimize latency over an extended period. The collector tries to perform only generation 0, generation 1, and concurrent generation 2 collections. Full blocking collections may still occur if the system is under memory pressure.)

2) Every second as a background task:
```
GCSettings.LargeObjectHeapCompactionMode = GCLargeObjectHeapCompactionMode.CompactOnce;
```
This is to ensure full blocking Gen2 GC's are also cleaning up the large objects.
(The large object heap (LOH) will be compacted during the next blocking generation 2 garbage collection.)

**This did not work well. Even under memory pressure full blocking  Gen2 GC's did not always happen in time and the app exceeded the memory limits.**

### Test B:

1) One time setting at startup:~~
```
GCSettings.LatencyMode = GCLatencyMode.Interactive;
```
This is the default value by .NET core 3.1 anyway.

2) Every second as a background task:
```
GCSettings.LargeObjectHeapCompactionMode = GCLargeObjectHeapCompactionMode.CompactOnce;
```
This is to ensure full blocking Gen2 GC's are also cleaning up the large objects.
(The large object heap (LOH) will be compacted during the next blocking generation 2 garbage collection.)

**This did not work well. Even under memory pressure full blocking  Gen2 GC's did not always happen in time and the app exceeded the memory limits.**

Any other ideas what I can try to get rid of the manual GC's?

**Comment by jkotas** [comment, marked as answer]
You can try setting HeapHardLimit or HeapHardLimitPercent: https://docs.microsoft.com/en-us/dotnet/core/run-time-config/garbage-collector#heap-limit-percent. It will tell GC to stay under the given limit.

Also, you can try turning off server GC by setting `<ServerGarbageCollection>false</ServerGarbageCollection>`. The server GC has better throughput, but it also needs more slack space to do a good job.

**Reply by JMan7777** [reply]
> 
> 
> You can try setting HeapHardLimit or HeapHardLimitPercent: https://docs.microsoft.com/en-us/dotnet/core/run-time-config/garbage-collector#heap-limit-percent. It will tell GC to stay under the given limit.
> 
> Also, you can try turning off server GC by setting `<ServerGarbageCollection>false</ServerGarbageCollection>`. The server GC has better throughput, but it also needs more slack space to do a good job.

@jkotas 
Thx a lot.  My Azure instance has 7.5 GB memory but the OS part in my container is using up to round 2.5 GB. That leaves me with maximum 5 GB left for my single instance app. 

I will give it a test with `COMPlus_GCHeapHardLimitPercent` set to 60% (3C in Hex) and will let you know how it goes.

**Reply by JMan7777** [reply]
### Update on testing COMPlus_GCHeapHardLimitPercent

I was now testing with a few COMPlus_GCHeapHardLimitPercent values and it seems that it solves my issue. 👍 

However I'm a bit confused.

**The MS documentation states:**
```
The default value, which only applies in certain cases, is the lesser of 20 MB or 75% of the memory limit on the container. The default value applies if:
    - The process is running inside a container that has a specified memory limit.
    - System.GC.HeapHardLimit is not set.
```

I was playing with certain values on my Azure web app (Linux docker container) and according to the documentation the default value should apply (which is 75%).

But it seems that's not the case. I was testing with setting COMPlus_GCHeapHardLimitPercent explicitly to 75% (Hex 4B).
With this value the app never exceeded 91% of the available memory.
I tested further and I could set the COMPlus_GCHeapHardLimitPercent up to 80% and it still did not reach the memory limit.

How can this be? 
Without setting the COMPlus_GCHeapHardLimitPercent limit explicitly the app will exceed the memory limit and crashes as it does not do GC's in time but when I set  COMPlus_GCHeapHardLimitPercent  to even 80% it's just fine.

Another question I have is, why there is no GCSettings.LargeObjectHeapCompactionMode for compact always (GCLargeObjectHeapCompactionMode.CompactAlways) ? 
I assume I'm not the only person who has to cleanup the LOH on regular bases.
Having the need to permanently ensuring GCSettings.LargeObjectHeapCompactionMode is set to GCLargeObjectHeapCompactionMode.CompactOnce is far from elegant. 
So could this be a future improvement feature for .NET core 6 ?

Thx a lot.

**Reply by jkotas** [reply]
The default GCHeap limits only kick in when the container memory is constrained (e.g. using `--memory` switch for docker). Is it the case for your docker container?

**Reply by jkotas** [reply]
> why there is no GCSettings.LargeObjectHeapCompactionMode for compact always

@Maoni0 ?

**Reply by JMan7777** [reply]
> 
> 
> The default GCHeap limits only kick in when the container memory is constrained (e.g. using `--memory` switch for docker). Is it the case for your docker container?

@jkotas 

I see in the logs that Azure is starting the container as follows (replaced few names with XXXXXXX):

```
docker run -d -p 6275:80 --name XXXXXXX -e DOCKER_CUSTOM_IMAGE_NAME=XXXXXXX:latest -e WEBSITE_SITE_NAME=XXXXXXX -e WEBSITE_AUTH_ENABLED=False -e PORT=80 -e WEBSITE_ROLE_INSTANCE_ID=0 -e WEBSITE_HOSTNAME=XXXXXXX.azurewebsites.net -e WEBSITE_INSTANCE_ID=XXXXXXX -e HTTP_LOGGING_ENABLED=1 XXXXXXX.azurecr.io/XXXXXXX:latest 
```

So I don't see that Azure is adding the `--memory` switch. 

As the limit of my B3 instance is 7.5 GB shouldn't MS add this flag automatically? I don't think that this is something I should be setting since instances can be up- and down scaled and explicit settings might get forgotten and not match the instance.
At least if the app instances number is set to 1 (like in my case, without auto scaling) the memory switch should be set automatically by Azure.

Thx again.

**Reply by jkotas** [reply]
> As the limit of my B3 instance is 7.5 GB shouldn't MS add this flag automatically?

@richlander Do you know somebody who can comment on this?

**Reply by Maoni0** [reply]
LOH is already automatically compacted in a constrained environment. 

when you got OOM what was the GC heap size? and when you say it's used 90% what was the GC heap size?

**Reply by JMan7777** [reply]
> 
> 
> LOH is already automatically compacted in a constrained environment.
> 
> when you got OOM what was the GC heap size? and when you say it's used 90% what was the GC heap size?

I did not get OOM originally (without any COMPlus_GCHeapHardLimitPercent  settings). 
The app just crashed and the Azure logs showed "Exceeding memory limit".

Only when I was setting COMPlus_GCHeapHardLimitPercent explicitly to a lower percentage e.g. 60% I could see OOM errors in the app logs.
Now having set COMPlus_GCHeapHardLimitPercent  to 80% my app maxes out around 96% of the available memory (7.5GB) and runs stable.

Do I understand you correct that I might not need to set 
`GCSettings.LargeObjectHeapCompactionMode = GCLargeObjectHeapCompactionMode.CompactOnce;` 
permanently in order to ensure that the LOH is cleaned up ?  If so, I can test this as well.

Thx

**Reply by Maoni0** [reply]
> Only when I was setting COMPlus_GCHeapHardLimitPercent explicitly to a lower percentage e.g. 60% I could see OOM errors in the app logs.
> Now having set COMPlus_GCHeapHardLimitPercent to 80% my app maxes out around 96% of the available memory (7.5GB) and runs stable.

to diagnose that we'd need to collect some data about the GC heap sizes. please see [this](https://github.com/Maoni0/mem-doc/blob/master/doc/.NETMemoryPerformanceAnalysis.md#how-to-collect-top-level-gc-metrics) for instructions if you need.

> Do I understand you correct that I might not need to set
GCSettings.LargeObjectHeapCompactionMode = GCLargeObjectHeapCompactionMode.CompactOnce;
permanently in order to ensure that the LOH is cleaned up ?

correct.

**Reply by JMan7777** [reply]
> > Do I understand you correct that I might not need to set
> > GCSettings.LargeObjectHeapCompactionMode = GCLargeObjectHeapCompactionMode.CompactOnce;
> > permanently in order to ensure that the LOH is cleaned up ?
> 
> correct.

Thx a lot. I will first try out this one and then reply. 👍

**Reply by JMan7777** [reply]
@jkotas @Maoni0 

**I tested the following scenario and it works just fine when I have only a single app instance running. 🥇 Thx a lot.  :**

- Only specify the COMPlus_GCHeapHardLimitPercent value to limit the memory usage of an app instance.

**_Still one question remains for me:_**

My Azure instance has 7.5GB memory available.
Inside this Azure instance i can scale or auto-scale 1-3 app instances.
This means the memory available to each app instance might change at any point in time depending on how many app instances are running.

1. How does .NET core know how many memory is available / free inside each app instance ?
2. How can I ensure that at any point in time for any of my app instances I do not run in the situation that .NET core tries to allocate more memory than available in the Azure instance.

**_Example:_**

_Azure instance with 10GB memory:_

10:00am - 1 app instance started using 2GB and having max 8GB GB free memory.
10:01am - Azure auto-scales a second app instance to start. The second app instance also uses 2GB which means that both app instances 1 & 2 now having together max 6GB GB free memory.
10:02am - Azure auto scales a third app instance to start. The third app instance also uses 2 GB which means that app instances 1 & 2 & 3 now having together max 4GB GB free memory.
10:03am - App 2 increases its memory usage to 3 GB which means that app instances 1 & 2 & 3 now having together max 3GB GB free memory.
10:04am - Azure stops the third app instance releasing 2 GB memory. App instances 1 & 2 now having together max 5GB GB free memory.

So when I specify COMPlus_GCHeapHardLimitPercent e.g. 50% what will the Garbage Collector calculate for each time and each instance?
Is the Garbage Collector so intelligent to basically adjust permanently the GCHeapHardLimit according to available memory on the system ?

10:00am  
- app 1 => 2GB => 10GB total memory, no other apps => GCHeapHardLimit = 5GB (50% of 10GB)

10:01am  
- app 1 => 2GB => 10GB total memory, app 2 using 2 GB => GCHeapHardLimit = 4GB (50% of 8GB)
- app 2 => 2GB => 10GB total memory, app 1 using 2 GB => GCHeapHardLimit = 4GB (50% of 8GB)

10:02am  
- app 1 => 2GB => 10GB total memory, app 2 using 2 GB, app 3 using 2 GB => GCHeapHardLimit = 3GB (50% of 6GB)
- app 2 => 2GB => 10GB total memory, app 1 using 2 GB, app 3 using 2 GB => GCHeapHardLimit = 3GB (50% of 6GB)
- app 3 => 2GB => 10GB total memory, app 1 using 2 GB, app 2 using 2 GB => GCHeapHardLimit = 3GB (50% of 6GB)

10:03am  
- app 1 => 2GB => 10GB total memory, app 2 using 3 GB, app 3 using 2 GB => GCHeapHardLimit = 2.5GB (50% of 5GB)
- app 2 => 3GB => 10GB total memory, app 1 using 2 GB, app 3 using 2 GB => GCHeapHardLimit = 2.5GB (50% of 5GB) => Memory usage over the limit.
- app 3 => 2GB => 10GB total memory, app 1 using 2 GB, app 2 using 3 GB => GCHeapHardLimit = 2.5GB (50% of 5GB)

10:04am  
- app 1 => 2GB => 10GB total memory, app 2 using 3 GB => GCHeapHardLimit = 3.5GB (50% of 7GB)
- app 2 => 3GB => 10GB total memory, app 1 using 2 GB => GCHeapHardLimit = 3.5GB (50% of 7GB)

Thx again.

**Reply by jkotas** [reply]
> So when I specify COMPlus_GCHeapHardLimitPercent e.g. 50% what will the Garbage Collector calculate for each time and each instance?

The hard limit percent is always computed from all available memory that the container can use. You can see the hard limit that is used by the process by logging value of `GC.GetGCMemoryInfo().TotalAvailableMemoryBytes`.

> Is the Garbage Collector so intelligent to basically adjust permanently the GCHeapHardLimit according to available memory on the system ?

It is not. The limit is computed at process start and does not change after that.

> How can I ensure that at any point in time for any of my app instances I do not run in the situation that .NET core tries to allocate more memory than available in the Azure instance.

Partition the memory between the app instances and give each instance enough for it to run.

I assume that you are using Azure Kubernetes Service. Is that right? https://kubernetes.io/docs/tasks/administer-cluster/manage-resources/memory-default-namespace/ is documentation for memory configuration in K8.

**Reply by JMan7777** [reply]
@jkotas 

Thx a lot for the great explanation. I was already assuming the Garbage Collector is behaving like that.

I'm not yet using Azure Kubernetes Service but planning to migrate in the near future.
As I'm for now running a single app instance I will just stick with setting a `COMPlus_GCHeapHardLimitPercent` value.

It's a bit of a a pitty that that each app instance need to be limited at startup and the available memory needs to be partitioned between the app instances. When you want to use app auto-scaling you waste a lot of memory which you can't use even it's free.

Thx again for the great help.

**Reply by Maoni0** [reply]
> It's a bit of a a pitty that that each app instance need to be limited at startup and the available memory needs to be partitioned between the app instances. When you want to use app auto-scaling you waste a lot of memory which you can't use even it's free.

can you please tell me what your desired behavior would be? if you have 10GB and start running 1 instance, do you want it to take up the whole 10GB? and if you start running a second instance, would you want each one to take 5GB? what if now the first instance's live data size is already 10GB which means it cannot go back to 5GB? 

I've discussed an API for folks to communicate a new limit to GC (so that GC doesn't need to query the limit and miss the mark when it changes) - it's trivial to communicate it to the GC; the non trivial part is to decide on a policy that makes the behavior reasonable for most apps when the new limited is communicated.

**Reply by JMan7777** [reply]
@Maoni0 

From my perspective the following solution would be great in order to make the best use of memory in today's containerized environments.

- Add two new GC settings which can be configured:
1. System.GC.HeapFreeHardLimit / COMPlus_GCHeapFreeHardLimit
2. System.GC.HeapFreeHardLimitPercent / COMPlus_GCHeapFreeHardLimitPercent

**System.GC.HeapFreeHardLimit / COMPlus_GCHeapFreeHardLimit**
Specifies the minimum free memory of the total physical memory in bytes at any point in time.

**System.GC.HeapFreeHardLimitPercent / COMPlus_GCHeapFreeHardLimitPercent**
Specifies the minimum free memory as a percentage of the total physical memory at any point in time.
If System.GC.HeapFreeHardLimit is also set, this setting is ignored.

These new settings are not exclusive to the existing GC settings and could be used even in combination with the existing settings.

**Example on 10GB system:**

If a developer wants a single app instance to never use more than 8GB and wants to keep always 2GB system memory free currently he would specify either e.g.:
- System.GC.HeapHardLimit / COMPlus_GCHeapHardLimit  = 8GB
- System.GC.HeapHardLimitPercent / COMPlus_GCHeapHardLimitPercent = 80%

However this is giving trouble and will crash app instances (most likely without given any OOM warning inside the app like in my case) when multiple of these app instances are started.

So when the developer adds (on top of the existing settings) either of the following  2 new settings e.g:
- System.GC.HeapFreeHardLimit / COMPlus_GCHeapFreeHardLimit = 2GB
- System.GC.HeapFreeHardLimitPercent / COMPlus_GCHeapFreeHardLimitPercent = 20%

.NET Core would ensure that that the total system memory is used as efficient as possible while still honoring all the set limits.
If the app instances still run in OOM situations a clear OOM error message could be thrown indicating which GC settings limit was exceeded. That way the developer could adjust GC settings or the max app instance count.

Additionally there could be maybe a `GC.GetGCMemoryInfo().TotalFreeMemoryBytes` method which returns back how much free memory is still left inside the app instance before reaching current OOM conditions. Together with the 2 new GC settings described above that would be my dream solution ;).

Thx again.

**Reply by Maoni0** [reply]
what does COMPlus_GCHeapFreeHardLimit mean? GC is per process. is this FreeHardLimit supposed to be the free limit of this process? if you mean it's a free limit of a container where multiple processes are running then it's unclear what each process is supposed to use. all GC knows is the working set of the process and what the memory limit is in the container. on Windows the memory load in a container is not exposed. having GC constantly enumerating all the processes within the same container would be quite expensive which was why I proposed a way to communicate a new limit to the GC.

> (most likely without given any OOM warning inside the app like in my case

I wouldn't attribute this to the GC without debugging, it may or may not have anything to do with the GC heap. it could just be that it's so short on memory some native code is having problems.

**Reply by Maoni0** [reply]
I dunno if cgroups provides a way to know how much memory other processes are using in the container. @janvorli do you know? if so we could definitely take advantage of it.

**Reply by JMan7777** [reply]
> what does COMPlus_GCHeapFreeHardLimit mean?

When currently .NET core processes / app instances are starting they somehow getting the total available memory:

System.GC.HeapHardLimitPercent / COMPlus_GCHeapHardLimitPercent = 80% in a 10GB system will set the COMPlus_GCHeapHardLimit to 8GB.

When running on a non-containerized environment .NET core should be also able to ask the OS about the total free memory.
When running on a containerized environment these free memory metrics should be available in the cgroup filesystem inside each container. 
According to: [https://docs.docker.com/config/containers/runmetrics](https://docs.docker.com/config/containers/runmetrics)

So when .NET core is able the get the free memory then COMPlus_GCHeapFreeHardLimit  can be used to check if an app instance is still having enough free memory available or is running in a OOM condition based on the COMPlus_GCHeapFreeHardLimit settings .

Regarding:

>    (most likely without given any OOM warning inside the app like in my case)

When I wasn't  specifying the COMPlus_GCHeapHardLimitPercent value, my  .NET core app instance never showed any OOM error. Azure just killed the app with the Azure error "Exceeding memory limits". However when I specified COMPlus_GCHeapHardLimitPercent and set it to a low value I saw OOM errors which are normal as I set the memory restrictions too tight.
I also was surprised that that without COMPlus_GCHeapHardLimitPercent settings .NET Core never gives any OOM warnings. It looks it just requests more memory than the Azure instance is subscribed to and Azure just kills in that case the app instance.

**Reply by Maoni0** [reply]
every process can read how much free space there is but they can't all use it or you would overcommit. if we can get the equivalent of the memory load of the OS for the container it'll make things better but it's not foolproof, just like running multiple processes on the same machine. the difference is just that folks tend to be a lot more mindful when it comes to containers.

**Reply by JMan7777** [reply]
> if we can get the equivalent of the memory load of the OS for the container it'll make things better but it's not foolproof,

It would be at least better than nothing. At the moment there is no good way available to have auto-scaled app instances available in a container environment without wasting memory. Since memory isn't cheap in Azure I was just hoping there could be some way to make better use of the memory. 

If you find a way equivalent of the memory load of the OS for the container, even if not foolproof it would be appreciated by many. Maybe introduce then related new GC settings as an experimental feature. ;)

**Reply by Maoni0** [reply]
as I mentioned above, we'll definitely take advantage of it if there's a way (I'm chatting with @janvorli about it) :) thanks for bringing this up!

**Reply by JMan7777** [reply]
You are welcome. Now my problem becomes your problem ;). hahaha
